### ISSUE TITLE
Allow using `rustls` instead of OpenSSL

### ISSUE THREAD

**Issue by Shnatsel** [frewsxcv/rust-crates-index#88]
In `cargo audit` we're looking to get rid of OpenSSL because of its security track record. So we need `crates-index` to allow using `rustls` instead of OpenSSL.

There is a number of way to accomplish this:

1. Allow using the out-of-tree `hyper` transport for `libgit2`: https://github.com/henry40408/git2-hyper
1. Switch from `libgit2` to `gitoxide`: https://github.com/frewsxcv/rust-crates-index/issues/82
1. https://github.com/frewsxcv/rust-crates-index/issues/87 (still being tested in Cargo)

Which option do you prefer, as the maintainer?

**Comment by kornelski** [comment]
My preferred solution would be 2. 

The git2-hyper solution seems technically reasonable, but the crate seems very young and already has an out-of-date git2 dependency, so this would need a more mature, actively-maintained implementation.

I don't mind adding a sparse registry support separately, but this only covers fetch of individual crates by name, and not listing them all and (currently WIP) learning about newly-published crates.

**Comment by Shnatsel** [comment]
Is migrating to `gitoxide` and completely dropping support for `libgit2` acceptable? 

It's technically possible to support both backends, but would create additional complexity and maintenance burden, so I'd prefer to avoid that.

**Comment by kornelski** [comment]
Yes

**Comment by BlackHoleFox** [comment]
Gonna give this a try and see if `git-repository` can fill in everything `git2` was doing.

**Comment by Shnatsel** [comment]
The [month-old progress update for `gitoxide`](https://github.com/Byron/gitoxide/discussions/694) details the ongoing integration into Cargo and the fixes that resulted from it. Those progress updates are [intentionally delayed](https://www.reddit.com/r/rust/comments/10if838/gitoxide_in_december_2022_the_cargo_integration/j5e1vln/) and the actual status is even better now.

@BlackHoleFox did you make any progress on migrating this crate to `gitoxide`? Do you think you could get it over the finish line in the next month or so?

**Comment by BlackHoleFox** [comment]
@Shnatsel Yeah, i'll pick this up again this week, sorry. I got thrown off by the amount of boilerplate you need to wrap `gitoxide` in to do common stuff (but maybe I'm going about it wrong, we'll see :) ) and then got distracted elsewhere.

**Comment by Shnatsel** [comment]
Alright! Perhaps you could check with Byron, the author of `gitoxide`, if you're on the right track and using the right APIs. Or perhaps this is something that could be improved in `gitoxide`!

**Comment by Shnatsel** [comment]
The integration into Cargo might also be useful as a reference point: https://github.com/rust-lang/cargo/pull/11448

**Comment by Byron** [comment]
> Alright! Perhaps you could check with Byron, the author of `gitoxide`, if you're on the right track and using the right APIs. Or perhaps this is something that could be improved in `gitoxide`!

Definitely! A cumbersome API is something to improve, especially if `git2` allows the same in an easier fashion. It's also advisable to track the latest `git-repository` releases as they typically come with improvements here and there.

**Comment by obi1kenobi** [comment]
Just a little +1 for this idea from the maintainers of the `cargo-semver-checks` crate. We've been dealing with obnoxious OpenSSL-related errors both when vendoring it in (e.g. https://github.com/obi1kenobi/cargo-semver-checks/issues/272) and when relying on it being dynamically-linked (the installed OpenSSL isn't always found, different systems require different solutions, and the whole thing just doesn't scale from an end user support perspective).

If/when `gitoxide` and `rustls` are able to be used instead of `git2` and OpenSSL, we'll probably switch to them quite quickly.

**Comment by Byron** [comment]
Apologies, this issue was auto-closed which wasn't actually my intention nor my doing even though GitHub claims it was 😅. Could you recheck to see if the issue is fixed with the latest version 2.0 or above? Thank you.

**Comment by complexspaces** [comment]
I don't believe that this is fully addressed yet. The new `gix` integration-based implementation is a start but due to [this line](https://github.com/frewsxcv/rust-crates-index/blob/master/Cargo.toml#L86) unconditionally selecting the "default" `curl` backend will pull in OpenSSL and build it. 

I think another feature would need to be added which exposes the `http-client-reqwest-rust-tls` flag in `gix-transport` in order to remove the OpenSSL dependency. `http-client-curl-rust-tls` might work as well but I imagine people looking for a "pure Rust" implementation don't want `curl` either, but I could be wrong!

**Comment by Byron** [comment]
It's true - to make this work one would have to prevent ending up with the `curl` feature in the first place, so the respective cargo features would have to be exposed in this crate to give a choice. A PR would definitely be welcome.
